### PR TITLE
feat: private parameters for application command

### PR DESCRIPTION
**Pull request by Dysta**
## Summary

When we want to call the callback of a command from another, it is interesting to be able to pass additional parameters which are only internal to the code. Currently we can't do this without performing a "trick" because each parameter indicated in the signature of a command is sent to discord, which allows the user to be able to manipulate.

The current trick consists of adding a **kwargs in the command signature and retrieving the parameters that interest us via a get. However, there is no guarantee that this trick will one day be corrected or will evolve and break the entire code.

This PR add a new type to slash commands that is called Private and it will not be registered to Discord.

## Checklist

- [x] If code changes were made, then they have been tested
    - [ ] I have updated the documentation to reflect the changes
    - [ ] I have formatted the code properly by running `task lint`
    - [ ] I have type-checked the code by running `task pyright`
- [ ] This PR fixes an issue
- [x] This PR adds something new (e.g. new method or parameters)
- [ ] This PR is a breaking change (e.g. methods or parameters removed/renamed)
- [ ] This PR is **not** a code change (e.g. documentation, README, ...)

### test code
```py
# in main.py
from secret import BOT_TOKEN

import logging
from disnake.ext import commands

logging.basicConfig(level=logging.DEBUG)

bot = commands.InteractionBot(
    test_guilds=[492677155180904448],
)

bot.load_extension(name="simple_cog")

# numb shouldn't be on discord
@bot.slash_command()
async def simple_slash_opt_tpe(inter, numb: commands.Private[int] = 87, letter: str = "letter"):
    await inter.response.send_message(f"Helo {numb} {letter}")


bot.run(BOT_TOKEN)

####################################

# in simple_cog.py
from disnake.ext import commands


class Simple(commands.Cog):
    def __init__(self, bot):
        self._bot = bot

    # let shouldn't be on discord
    @commands.slash_command()
    async def simple_cog_slash_opt(self, inter, numb: int = 5, let: commands.Private[str] = "Private"):
        await inter.response.send_message(f"Helo cog {numb} {let}")


def setup(bot):
    bot.add_cog(Simple(bot))
```

### CODE REVIEW

**Comment by onerandomusername** [comment]
I like the idea, @Dysta, are you able to rebase and solve conflicts with the master branch, plus fix checks?

Sorry for the delayed response, but the internal blocker for this has been pushed back so it is possible to add this now.

**Comment by Enegg** [comment]
> * First and foremost, the act of implicitly marking params starting with an `_` as private. This will break a large number of commands, as people often use e.g. `_with` instead of `with` to get around it being a built-in; the same goes for other builtins.

Iirc PEP says that to avoid name collisions, one should add the `_` at the end of the name, not in the front, so yeah people be breaking conventions

**Comment by Chromosomologist** [comment]
Mhm, that's true, but help channel activity tells me people often do otherwise. I don't necessarily mind the feature, but I'm not too keen on how breaking it is. 

Perhaps we could do so in an inevitable slash refactor or a major version change?

**Comment by onerandomusername** [comment]
@Dysta would you mind if I clean up the history on this pr a bit?

**Comment by Dysta** [comment]
> @Dysta would you mind if I clean up the history on this pr a bit?

you can, the history came bc I rebase the branch with the current master

File: disnake/ext/commands/params.py [hunk 1]
```diff
@@ -333,6 +336,23 @@ class LargeInt(int):
 _VERIFY_TYPES: Final[FrozenSet[OptionType]] = frozenset((OptionType.user, OptionType.mentionable))
 
 
+class _PrivateEnum(enum.Enum):
+    # An enum to generate sentinel values for Annotated
+    PRIVATE = enum.auto()
```
Comment (onerandomusername) [inline disnake/ext/commands/params.py:341]: what exactly is this used for? It doesn't seem like its used at runtime

Comment (Chromosomologist) [reply disnake/ext/commands/params.py:341]: The sentinel value is only relevant for runtime checking, in this case it really doesn't matter what metadata you attach to the `Annotated`, as it's not used anyways.

**Pull request closed without merging**